### PR TITLE
scheduler: support priority for each dimension in hot scheduling

### PR DESCRIPTION
**Pull request by Yisaer**
Signed-off-by: yisaer <disxiaofei@163.com>

### What is changed and how it works?

support priority for each dimension in hot scheduling

### Check List

Tests

- Unit test

I used `oltp_read_only` to test the dimension priority in hot scheduling, here is the result.

image

### Release note

```release-note
Support priority for each dimension in hot scheduleing
```

### CODE REVIEW

File: server/schedulers/hot_region.go [hunk 1]
```diff
@@ -1212,3 +1251,24 @@ func getRegionStatKind(rwTy rwType, dim int) statistics.RegionStatKind {
 	}
 	return 0
 }
+
+func (bs *balanceSolver) preferPriority() [2]string {
+	priorities := bs.sche.conf.WritePriorities
+	if bs.rwTy == read {
+		priorities = bs.sche.conf.ReadPriorities
+	}
+	if len(priorities) == 1 {
```
Comment (lhy1024) [inline server/schedulers/hot_region.go:1260]: when will it happen? do we need to consider that len is zero?

File: server/schedulers/hot_region.go [hunk 2]
```diff
@@ -773,37 +787,19 @@ func (bs *balanceSolver) calcProgressiveRank() {
...
-			dstRate := dstLd.Loads[dim]
-			peerRate := peer.GetLoad(getRegionStatKind(bs.rwTy, dim))
-			decRatio := (dstRate + peerRate) / getSrcDecRate(srcRate, peerRate)
-			isHot := peerRate >= bs.getMinRate(dim)
-			return isHot, decRatio
-		}
-		keyHot, keyDecRatio := checkHot(statistics.KeyDim)
-		byteHot, byteDecRatio := checkHot(statistics.ByteDim)
-
+		firstPriorityDimHot, firstPriorityDecRatio, secondPriorityDimHot, secondPriorityDecRatio := bs.getHotDecRatioByPriorities()
```
Comment (lhy1024) [inline server/schedulers/hot_region.go:790]:
```suggestion
		firstPriorityDimHot, firstPriorityDecRatio, secondPriorityDimHot, secondPriorityDecRatio := bs.getHotDecRatioByPriorities(srcLd, dstLd,peer)
```

File: server/schedulers/hot_region.go [hunk 3]
```diff
@@ -911,14 +936,13 @@ func (bs *balanceSolver) compareSrcStore(st1, st2 uint64) int {
 				)),
 			)
 		} else {
+			firstLdRankCmp, secondLdRankCmp, firstLdRankCmpZeroStep := bs.getLdRankPriorities()
 			lpCmp = sliceLPCmp(
 				minLPCmp(negLoadCmp(sliceLoadCmp(
-					stLdRankCmp(stLdByteRate, stepRank(bs.maxSrc.Loads[statistics.ByteDim], bs.rankStep.Loads[statistics.ByteDim])),
-					stLdRankCmp(stLdKeyRate, stepRank(bs.maxSrc.Loads[statistics.KeyDim], bs.rankStep.Loads[statistics.KeyDim])),
+					firstLdRankCmp,
+					secondLdRankCmp,
```
Comment (lhy1024) [inline server/schedulers/hot_region.go:943]:
```suggestion
stLdRankCmp(stLdByteRate, stepRank(bs.maxSrc.Loads[bs.firstPriority], bs.rankStep.Loads[bs.firstPriority])),
					stLdRankCmp(stLdKeyRate, stepRank(bs.maxSrc.Loads[bs.secondPriority], bs.rankStep.Loads[bs.secondPriority])),
```

File: server/schedulers/hot_region.go [hunk 4]
```diff
@@ -893,6 +911,13 @@ func (bs *balanceSolver) betterThan(old *solution) bool {
 	return false
 }
 
+func (bs *balanceSolver) getRkCmpPriorities(old *solution) (firstCmp int, secondCmp int) {
+	fk, sk := getRegionStatKind(bs.rwTy, bs.firstPriority), getRegionStatKind(bs.rwTy, bs.secondPriority)
+	fkRkCmp := rankCmp(bs.cur.srcPeerStat.GetLoad(fk), old.srcPeerStat.GetLoad(fk), stepRank(0, 100))
+	skRkCmp := rankCmp(bs.cur.srcPeerStat.GetLoad(sk), old.srcPeerStat.GetLoad(sk), stepRank(0, 10))
```
Comment (lhy1024) [inline server/schedulers/hot_region.go:917]: In general, we use 100 as the step rank of byte dim and 10 as the step rank of key dim.

File: server/schedulers/hot_region.go (hunk 1 above)
Comment (Yisaer) [reply server/schedulers/hot_region.go]: updated.

**Review by lhy1024 — APPROVED** [review]

**Review by nolouch — APPROVED** [review]

**Pull request merged**